### PR TITLE
♻️ Enable passing type-checking on src/experiments and src/examiner

### PR DESCRIPTION
**Pull request by rcebulko**
Part of #34096

### CODE REVIEW

File: src/examiner/examiner.js [hunk 1]
```diff
@@ -18,21 +18,20 @@
  * @param {!Window} win
  */
 function detectLongTasks(win) {
-  const observer = new win.PerformanceObserver(function (entryList) {
+  const observer = new win.PerformanceObserver((entryList) => {
+    /** @type {!Array<!PerformanceLongTaskTiming>} */
```
Comment (rcebulko) [inline src/examiner/examiner.js:22]: The observer is only observing `'long-tasks'`. This type-cast allows access to the `attribution` property, which is not present on all `PerformanceEntry`s

File: src/examiner/examiner.js [hunk 2]
```diff
@@ -18,21 +18,20 @@
...
         continue;
       }
-      const attr = entries[i].attribution[0];
-      if (!attr || !attr.containerSrc) {
+      const attr = attribution[0];
+      if (!attr?.containerSrc) {
         continue;
       }
 
-      const {duration} = entries[i];
```
Comment (samouri) [inline src/examiner/examiner.js:35]: this was just completely unused?

File: src/experiments/index.js [hunk 3]
```diff
@@ -21,11 +21,15 @@
...
-import {parseQueryString} from './url';
+import {dev, user} from '../log';
+import {getMode} from '../mode';
+import {getTopWindow} from '../service';
+import {hasOwn, map} from '../core/types/object';
+import {isArray} from '../core/types';
+import {parseQueryString} from '../url';
+
+// typedef imports
+import {ExperimentInfoDef} from './experiments.type';
```
Comment (samouri) [inline src/experiments/index.js:32]: this....this works??? I had no idea.

File: src/experiments/index.js [hunk 4]
```diff
@@ -142,42 +135,34 @@ export function experimentToggles(win) {
...
   Object.assign(toggles, getExperimentToggles(win));
 
-  if (
-    win.AMP_CONFIG &&
-    Array.isArray(win.AMP_CONFIG['allow-url-opt-in']) &&
-    win.AMP_CONFIG['allow-url-opt-in'].length > 0
-  ) {
-    const allowed = win.AMP_CONFIG['allow-url-opt-in'];
+  const allowedUrlOptIn = win.AMP_CONFIG?.['allow-url-opt-in'];
+  if (isArray(allowedUrlOptIn) && allowedUrlOptIn.length) {
```
Comment (samouri) [inline src/experiments/index.js:156]: nit: are there things other than arrays that have length? Can it be condensed to: `win.AMP_CONFIG['allow-url-opt-in']?.length`

File: src/experiments/story-ad-auto-advance.js [hunk 5]
```diff
@@ -25,29 +35,22 @@ export const StoryAdAutoAdvance = {
   TEN_SECONDS: '31060907',
 };
 
-/** @const Object<string, string> */
-export const AdvanceExpToTime = {
-  '31060905': '6s',
-  '31060906': '8s',
-  '31060907': '10s',
+/** @const @private {!ExperimentInfoDef} */
```
Comment (samouri) [inline src/experiments/story-ad-auto-advance.js:38]: is there a reason to hoist this? it isn't reused anywhere, I believe colocated with its usage was a better place to put it. 

nit: the name should also have `experiment` in it somewhere

File: src/examiner/examiner.js (hunk 2 above)
Comment (rcebulko) [reply src/examiner/examiner.js:35]: It's now destructured at the top of the loop

File: src/experiments/index.js (hunk 4 above)
Comment (rcebulko) [reply src/experiments/index.js:156]: If this property was assigned a string value, `length` could be truthy but for...of would have weird results

File: src/experiments/story-ad-auto-advance.js (hunk 5 above)
Comment (rcebulko) [reply src/experiments/story-ad-auto-advance.js]: Moved back into function

Comment (samouri) [reply src/experiments/story-ad-auto-advance.js]: thank you for doing this at both locations

**Review by samouri — APPROVED** [review]

**Review by calebcordry — APPROVED** [review]
a4a approval

**Review by rebeccanthomas — APPROVED** [review]

**Pull request merged**